src/realm/sync/subscriptions.cpp

```diff
@@ -463,25 +474,13 @@ util::Future<SubscriptionSet::State> SubscriptionSet::get_state_change_notificat
 {
     auto mgr = get_flx_subscription_store(); // Throws
 
-    std::unique_lock<std::mutex> lk(mgr->m_pending_notifications_mutex);
+    util::CheckedLockGuard lk(mgr->m_pending_notifications_mutex);
     // If we've already been superceded by another version getting completed, then we should skip registering
     // a notification because it may never fire.
     if (mgr->m_min_outstanding_version > version()) {
         return util::Future<State>::make_ready(State::Superseded);
     }
 
-    // Begin by blocking process_notifications from starting to fill futures. No matter the outcome, we'll
-    // unblock process_notifications() at the end of this function via the guard we construct below.
-    mgr->m_outstanding_requests++;
```

Comment:
It appears that the goal of this semaphore was to allow multiple concurrent calls to get_state_change_notification() to be refreshing at the same time. That doesn't seem like a very important thing to optimize for, and waiting on a condition variable is *much* more expensive than waiting on a mutex.